packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table.tsx

```diff
@@ -69,15 +71,22 @@ const getDataTypesOfPropertyType = (
   propertyType: PropertyType,
   entity: EntityResponse,
 ) => {
-  /** @todo check why propertyValue does not have with a proper type  */
-  return propertyType.oneOf.map((propertyValue: any) => {
-    if (propertyValue?.$ref) {
+  return propertyType.oneOf.map((propertyValue) => {
+    if ("$ref" in propertyValue) {
       const dataTypeId = propertyValue?.$ref;
-      return (
-        entity.entityTypeRootedSubgraph.referencedDataTypes.find(
-          (val) => val.dataTypeId === dataTypeId,
-        )?.dataType.title ?? "undefined"
-      );
+      const dataTypeVertex =
+        entity.entityTypeRootedSubgraph.vertices[dataTypeId];
+
+      if (!dataTypeVertex) {
+        return "undefined";
+      }
+      if (dataTypeVertex?.kind !== "dataType") {
+        throw new Error(
+          `Expected "dataType" vertex but got ${dataTypeVertex.kind}`,
+        );
+      }
```

Comment:
Not sure how we want to handle these sorts of things. I wrote this before I wrote the helper methods so we could use `isDataType` here,  but generally speaking I think we _should_ trust that a datatype ID will go to a data type? And just use type assertions?

Not sure, maybe we'd prefer more safety, or maybe we'd prefer it to be a bit faster and avoid the checks.

cc: @CiaranMn if you have any thoughts

Reply:
I expect we would need to have millions of these checks before it made any discernible difference to performance.

We can remove them if something appears slow and it would make any difference.

Reply:
Well yeah it's also about code ergonomics. Do we want to be writing the checks everywhere, or do we just want to write `as DataTypeVertex`

Reply:
I'm not against asserting the type if we're 100% sure it logically must we are asserting it as, but I don't think we should do so because of performance worries in this instance.

Reply:
> Do we want to be writing the checks everywhere, or do we just want to write as DataTypeVertex

Presumably there's only a limited number of ways in which we want to retrieve a thing from an array and check it is what it's supposed to do, so we should have a limited number of functions that do that, and keep the checks inside them.

I don't feel strongly about this, if I saw this and it had an assertion with a comment explaining that it was definitely going to be the thing I wouldn't care. But we should comment any assertion.

Reply:
Do we want to be adding a comment everywhere we do this:
https://github.com/hashintel/hash/blob/fb8ad4407a26e9cbd02f414f4a7e3c2439838553/packages/hash/frontend/src/pages/%5Baccount-slug%5D/entities/%5Bentity-id%5D.page/types-section.tsx#L43-L47

Or do we want to just create methods for all of them:
`getDataType(dataTypeId: VersionedUri, subgraph: Subgraph): PersistedDataType`
`getPropertyType(propertyTypeId: VersionedUri, subgraph: Subgraph): PersistedPropertyType`

and so on

Reply:
I would create methods for them. Again, I don't feel strongly about this, I'm offering an opinion because you've asked, feel free to think it's too cumbersome

Reply:
I've added them: https://github.com/hashintel/hash/pull/1220/commits/6dc0054a8d0d3ef622f0863b00a5f23ff3f7e810
And used them: https://github.com/hashintel/hash/pull/1220/commits/871caba9e52910aebdbddf22a61ae65644788dcb

Seem to improve ergonomics quite a bit. I'm wondering about naming, they'll clash a bit with BP hooks if we remove the `Persisted` prefix, (e.g. `getDataType`) so I'm wondering about maybe `extractDataType`?